code/datums/components/forensics.dm

```diff
@@ -0,0 +1,123 @@
+/datum/component/forensics
+	var/list/fingerprints
+	var/list/hiddenprints
+	var/list/blood_DNA
+	var/list/fibers
+	var/fingerprintslast
+
+/datum/component/forensics/proc/return_fingerprints()
+	return length(fingerprints)? fingerprints.Copy() : list()
+
+/datum/component/forensics/proc/return_hiddenprints()
+	return length(hiddenprints)? hiddenprints.Copy() : list()
+
+/datum/component/forensics/proc/return_blood_DNA()
+	return length(blood_DNA)? blood_DNA.Copy() : list()
+
+/datum/component/forensics/proc/return_fibers()
+	return length(fibers)? fibers.Copy() : list()
+
+/datum/component/forensics/proc/wipe_fingerprints()
+	fingerprints = null
+
+/datum/component/forensics/proc/wipe_hiddenprints()
+	return	//no.
+
+/datum/component/forensics/proc/wipe_blood_DNA()
+	blood_DNA = null
+
+/datum/component/forensics/proc/wipe_fibers()
+	fibers = null
+
+/datum/component/forensics/proc/add_fingerprints(list/_fingerprints)
+	if(!length(_fingerprints))
+		return
+	LAZYINITLIST(fingerprints)
+	for(var/i in _fingerprints)	//We use an associative list, make sure we don't just merge a non-associative list into ours.
+		fingerprints[i] = i
+
+/datum/component/forensics/proc/add_fingerprint_from_mob(mob/living/M, ignoregloves = FALSE)
+	if(!M)
+		return
+	add_hiddenprint(M)
+	if(ishuman(M))
+		var/mob/living/carbon/human/H = M
+		add_fibers(H)
+		if(H.gloves) //Check if the gloves (if any) hide fingerprints
+			var/obj/item/clothing/gloves/G = H.gloves
+			if(G.transfer_prints)
+				ignoregloves = 1
+			if(!ignoregloves)
+				H.gloves.add_fingerprint_from_mob(H, 1) //ignoregloves = 1 to avoid infinite loop.
+				return
+		var/full_print = md5(H.dna.uni_identity)
+		LAZYSET(fingerprints, full_print, full_print)
+
+/datum/component/forensics/proc/add_fibers(list/_fibertext)
+	if(!length(_fibertext))
+		return
+	LAZYINITLIST(fibers)
+	for(var/i in _fibertext)	//We use an associative list, make sure we don't just merge a non-associative list into ours.
+		fibers[i] = i
+
+/datum/component/forensics/proc/add_fibers_from_mob(mob/living/carbon/human/M)
+	var/fibertext
+	var/item_multiplier = isitem(src)?1.2:1
+	if(M.wear_suit)
+		fibertext = "Material from \a [M.wear_suit]."
+		if(prob(10*item_multiplier) && !(fibertext in fibers))
+			LAZYADD(fibers, fibertext)
+		if(!(M.wear_suit.body_parts_covered & CHEST))
+			if(M.w_uniform)
+				fibertext = "Fibers from \a [M.w_uniform]."
+				if(prob(12*item_multiplier) && !(fibertext in fibers)) //Wearing a suit means less of the uniform exposed.

```

Comment:
make fibers assoc.

Reply:
kk gimme a bit ill do it after i process blood dna (pain)